File: tutorials/python/ai_return_nowcaster.py
import os
import requests
import pandas as pd

# Define the API root URL
base_product_url = "https://monkfish-app-xcac2.ondigitalocean.app/"
primary_api_key = "needed-for-authentication"
secondary_api_key = "needed-for-authentication"

# AI RETURN NOWCASTER MODEL
model = "air_nowcaster"

endpoints = [
    "short_term_return_nowcast",
    "sdf_price_range",
    "nav_discount",
]

# Define the data to be sent in the request body (as a dictionary)
fund_segments = requests.get(os.path.join(base_product_url, "common/fund_segments"))
print("fund_segments", fund_segments.json())
macro_environments = requests.get(
    os.path.join(base_product_url, "common/macro_environments")
)
print("macro_environments", macro_environments.json())

request_body_short_term_return_nowcast = {
    "fund_segment": fund_segments.json()[0],
    "number_funds": 1,
}

request_body_sdf_price_range = {
    "list_input": [
        {"cash_flow_amount": 100, "exit_time": 1},
        {"cash_flow_amount": 150, "exit_time": 2},
    ],
    "fund_segment": fund_segments.json()[0],
    "macro_environment": macro_environments.json()[0],
    "rf_rate": 0,
}

request_body_nav_discount = {
    "fund_segment": fund_segments.json()[0],
    "nav": 100,
    "age": 10,
}


for endpoint in endpoints:
    # Build API URL
    url = os.path.join(base_product_url, model, endpoint)

    # Send the POST request
    if endpoint == "short_term_return_nowcast":
        request_body = request_body_short_term_return_nowcast
    elif endpoint == "sdf_price_range":
        request_body = request_body_sdf_price_range
    elif endpoint == "nav_discount":
        request_body = request_body_nav_discount
    else:
        raise ValueError(f"endpoint {endpoint} not defined.")

    # Set header for authentication
    headers = {"X-BLOBR-KEY": primary_api_key}

    # Send the POST request
    response = requests.post(url, json=request_body, headers=headers)

    if response.status_code == 401:
        # needed for API Key Rotation
        # More info: https://www.blobr.io/post/api-keys-best-practices
        headers = {"X-BLOBR-KEY": secondary_api_key}
        response = requests.post(url, json=request_body, headers=headers)

    # Check the response status code
    if response.status_code == 200:
        # Request was successful
        print(f"POST request successful: {url}")
        # print("Response JSON:", response.json())
    else:
        # Request failed
        print("POST request failed with status code:", response.status_code)

    # Convert dict to pd.DataFrame
    if endpoint == "sdf_price_range":
        df = pd.DataFrame(response.json(), index=["price_now"])
    elif endpoint == "nav_discount":
        df = pd.DataFrame(response.json(), index=["nav_now"])
    else:
        df = pd.DataFrame(response.json())
    # df.index.name = "horizon_in_years"
    # print("df", df.columns, df)

    # Save pd.DataFrame to .csv file
    df.to_csv(f"ai_return_nowcaster---{endpoint}.csv")
    # plot = df.plot(title="My Plot")

File: tutorials/R/ai_return_nowcaster.R
###########################################
# Plot AI Return Nowcaster Endpoint Results
###########################################
# Initialize ----
if(sys.nframe() == 0L) rm(list = ls())
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
getwd()

# install & load package "httr" to access API later
if (!require("httr", quietly = TRUE)) {
  install.packages("httr")
}

# Set API Base URL & API keys -----

base_product_url <- "https://base-product-url.app"
primary_api_key <- "needed-for-authentication"
secondary_api_key <- "needed-for-authentication"

# source("api_root.R")


# Send API requests ----
# Retrieve available fund_segments and macro_environments
post_request <- httr::GET(
  url = paste0(base_product_url, "common/fund_segments"),
  add_headers(.headers = c("X-BLOBR-KEY" = primary_api_key))
  )
fund_segments <- unlist(content(post_request, "parsed"))
print(fund_segments)
post_request <- httr::GET(url = paste0(base_product_url, "common/macro_environments"))
macro_environments <- unlist(content(post_request, "parsed"))
print(macro_environments)

# Choose one of these three endpoints
endpoint <- "air_nowcaster/short_term_return_nowcast"
endpoint <- "air_nowcaster/sdf_price_range"
endpoint <- "air_nowcaster/nav_discount"

download.air.nowcast <- function(endpoint) {
  
  # Define the request body
  if (endpoint == "air_nowcaster/short_term_return_nowcast") {
    request_body <- list(
      fund_segment = fund_segments[1],
      number_funds = 2
    )
  } else if (endpoint == "air_nowcaster/sdf_price_range") {
    request_body <- list(
      list_input = list(
        list(cash_flow_amount=100, exit_time=1), # example cash flow stream
        list(cash_flow_amount=150, exit_time=2), # example cash flow stream
        list(cash_flow_amount=220, exit_time=3) # example cash flow stream
      ),
      fund_segment= fund_segments[1],
      macro_environment= macro_environments[1],
      rf_rate= 0
    )
  } else if (endpoint == "air_nowcaster/nav_discount") {
    request_body = list(
      nav = 100, # current fund net asset value
      age = 10, # current fund age
      fund_segment = fund_segments[1]
    )
  }
  
  # Build API URL
  api_url <- paste0(base_product_url, endpoint)
  
  # Create the POST request
  post_request <- httr::POST(api_url,
                             add_headers(.headers = c("X-BLOBR-KEY" = primary_api_key)),
                             body = request_body,
                             encode = "json")
  print(post_request)
  
  
  # Send the POST Request:
  response <- httr::content(post_request, "parsed")
  print(response)
  
  # Convert to data.frame
  df <- data.frame(lapply(response, unlist))
  
  # Plot data.fame containing results
  if (endpoint == "air_nowcaster/short_term_return_nowcast") {
    quarter <- 4
    df.q <- df[1:(nrow(df)-1), ]
    x <- as.numeric(sub("quantile_", "", rownames(df.q)))
    y <- df.q[, paste0("quarter_", quarter)]
    plot(x, y, type="h", 
         xlab = "Quantiles", ylab = "Return", 
         main = paste("Cumulative return quantiles after", quarter, "quarter(s)")
         )
    abline(h = df["mean", paste0("quarter_", quarter)], col = "blue", lwd = 2, lty = 3)
    legend("bottomright", bty = "n", legend = "expected return", col = "blue", lwd = 2, lty = 3)
  } else if (endpoint == "air_nowcaster/sdf_price_range") {
    # Define mean and standard deviation
    m <- df$price_mean  # Mean
    s <- df$price_stdv  # Standard deviation
    
    # Generate x-values for the plot
    x <- seq(m - 4 * s, m + 4 * s, length = 1000)
    
    # Calculate the PDF values for the given mean and standard deviation
    pdf_values <- dnorm(x, mean = m, sd = s)
    
    # Create the plot
    plot(x, pdf_values, type = "l", col = "blue", lwd = 2,
         xlab = "Fund Price (discounted cash flows)", 
         ylab = "PDF", main = "Fund Price Range")
    abline(h=0, col="grey")
    abline(v=m, col="green", lwd=2)
    legend("topright", bty="n", legend = "expected price", col="green", lwd=2, lty=1)
  } else if (endpoint == "air_nowcaster/nav_discount") {
    barplot(as.matrix(df), 
            xlab ="Net Asset Value", 
            main = "NAV Discount Analysis",
            ylim = c(0, max(df) *1.1))
  }
  
  return(df)
}

df <- download.air.nowcast(endpoint = "air_nowcaster/short_term_return_nowcast")
df <- download.air.nowcast(endpoint = "air_nowcaster/sdf_price_range")
df <- download.air.nowcast(endpoint = "air_nowcaster/nav_discount")

